src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs

```diff
@@ -0,0 +1,373 @@
+
+using System.Collections.Immutable;
+
+namespace System.Reflection.Metadata.Decoding
+{
+    /// <summary>
+    /// Decodes custom attribute blobs.
+    /// </summary>
+    internal struct CustomAttributeDecoder<TType>
+    {
+        private ICustomAttributeTypeProvider<TType> _provider;
+        private MetadataReader _reader;
+
+        public CustomAttributeDecoder(ICustomAttributeTypeProvider<TType> provider, MetadataReader reader)
+        {
+            _reader = reader;
+            _provider = provider;
+        }
+
+        public CustomAttributeValue<TType> DecodeValue(EntityHandle constructor, BlobHandle value)
+        {
+            BlobHandle signature;
+            switch (constructor.Kind)
+            {
+                case HandleKind.MethodDefinition:
+                    MethodDefinition definition = _reader.GetMethodDefinition((MethodDefinitionHandle)constructor);
+                    signature = definition.Signature;
+                    break;
+
+                case HandleKind.MemberReference:
+                    MemberReference reference = _reader.GetMemberReference((MemberReferenceHandle)constructor);
+                    signature = reference.Signature;
+                    break;
+
+                default:
+                    throw new BadImageFormatException();
+            }
+
+            BlobReader signatureReader = _reader.GetBlobReader(signature);
+            BlobReader valueReader = _reader.GetBlobReader(value);
+
+            ushort prolog = valueReader.ReadUInt16();
+            if (prolog != 1)
+            {
+                throw new BadImageFormatException();
+            }
+
+            SignatureHeader header = signatureReader.ReadSignatureHeader();
+            if (header.Kind != SignatureKind.Method || header.IsGeneric)
+            {
+                throw new BadImageFormatException();
+            }
+
+            int parameterCount = signatureReader.ReadCompressedInteger();
+            SignatureTypeCode returnType = signatureReader.ReadSignatureTypeCode();
+            if (returnType != SignatureTypeCode.Void)
+            {
+                throw new BadImageFormatException();
+            }
+
+            ImmutableArray<CustomAttributeTypedArgument<TType>> fixedArguments = DecodeFixedArguments(ref signatureReader, ref valueReader, parameterCount);
+            ImmutableArray<CustomAttributeNamedArgument<TType>> namedArguments = DecodeNamedArguments(ref valueReader);
+            return new CustomAttributeValue<TType>(fixedArguments, namedArguments);
+        }
+
+        private ImmutableArray<CustomAttributeTypedArgument<TType>> DecodeFixedArguments(ref BlobReader signatureReader, ref BlobReader valueReader, int count)
+        {
+            if (count == 0)
+            {
+                return ImmutableArray<CustomAttributeTypedArgument<TType>>.Empty;
+            }
+
+            var arguments = new CustomAttributeTypedArgument<TType>[count];
```

Comment:
Maybe, there's a tradeoff between allocating small outer builder object vs. copying from pooled builder to final destination, plus any waste associated with unused pooled objects. I'm not sure which is better. For now, I went with same pattern as signature decoder (MoveToImmutable). If we should switch to pooled buildes, we should do it across the board.